net: lib: zperf: fix internal UDP header definition

JordanYates opened the pull request:
Fix the zperf UDP datagram header definition, which has had the following definition in the `iperf` repo since at least October 2019. https://sourceforge.net/p/iperf2/code/ci/master/tree/include/payloads.h#l165
```
struct UDP_datagram {
// used to reference the 4 byte ID number we place in UDP datagrams
// Support 64 bit seqno on machines that support them
    uint32_t id;
    uint32_t tv_sec;
    uint32_t tv_usec;
    uint32_t id2;
};
```
The response decoding was testd with a nRF7002 client and `iperf-2.2.1-win64.exe` server, with the output statistics struct now containing the same information as reported on the PC server.

Output a warning if the server response does not have the expected header flag set.

Powershell server:
![image](https://github.com/user-attachments/assets/71ee00f0-99dd-473e-9a61-21b737facd35)
Embedded decoding response:
![image](https://github.com/user-attachments/assets/ee96cc92-1d0f-4be7-af7f-9319ce585559)

JordanYates committed:
net: lib: zperf: fix internal UDP header definition

Fix the zperf UDP datagram header definition, as iperf introduced a
backwards incompatible format change in version 2.0.10 (August 2017).
```
struct UDP_datagram {
// used to reference the 4 byte ID number we place in UDP datagrams
// Support 64 bit seqno on machines that support them
    uint32_t id;
    uint32_t tv_sec;
    uint32_t tv_usec;
    uint32_t id2;
};
```

Update the header to the new format, with a Kconfig option to fall back
to the previous header definition.

The response decoding was testd with a nRF7002 client and
`iperf-2.2.1-win64.exe` server, with the output statistics struct now
containing the same information as reported on the PC server.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
net: lib: zperf: warning on bad response flags

Output a warning if the server response does not have the expected
header flag set.

Signed-off-by: Jordan Yates <jordan@embeint.com>

rlubos reviewed (commented):
LGTM, however we should probably update the documentation as well? We do mention in a cople of places that zperf is compatible with iPerf_2.0.5 which seems no longer be the case?
https://github.com/zephyrproject-rtos/zephyr/blob/90cd350c5a4a25ddee7886a9f58c8d0bc86db7da/doc/connectivity/networking/api/zperf.rst?plain=1#L16
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/zperf/README.rst?plain=1#L16

jukkar commented:
No objection to supporting iperf 2.2, but we need to keep old iperf version supported too. So I suggest adding a choice kconfig option that will allow user to select to use the newer 2.2 version too (and by default the support should be with older iperf as that we have been documenting people to use).

JordanYates commented:
> No objection to supporting iperf 2.2, but we need to keep old iperf version supported too. So I suggest adding a choice kconfig option that will allow user to select to use the newer 2.2 version too (and by default the support should be with older iperf as that we have been documenting people to use).

iperf 2.0.5 was released no later than March 2014 (The 2.0.6 release date), I don't agree that we should be defaulting to a version over a decade old.
Do you know when the header layout changed? I can only tell that it was over 5 years ago.
Which again, not sure we need to be supporting versions over 5 years old.

jukkar commented:
The mess with various iperf version is because zephyr iperf support was created when iperf 2.0.5 was active, and iperf devs screwed things up by not having proper versioning in the iperf protocol. So now we have a problem that zperf cannot know in advance what iperf2 version the other end is using.

JordanYates commented:
That doesn't solve any problem, it just shifts who is going to experience it.

If we go with your suggestion of the legacy default, any user that installs an iperf release from the last 5 years is going to have a non-working sample by default. I would argue that the number of people with an iperf installation older than 5 years old is dwarfed by the number of people with a newer version. Ubuntu 24.04 for example ships with `2.1.9` (March 2023).

And just as a general troubleshooting process, if I find a tool that doesn't work, my first instinct is to try the latest version, not roll back 6 years.

jukkar commented:
> That doesn't solve any problem, it just shifts who is going to experience it.

What is the problem we are trying to solve here? I am guessing you downloaded / used some latest iperf2 version that did not work, then looked why it didn't work and decided to fix zephyr side. That is fine too, but we have in our documentation mention of iperf version that works with zperf, and of course we can update the documentation together with zperf implementation to use newer iperf2.2 version. I am just wanting to avoid getting issues created because of this as this.

What we could do here is:
* allow user to choose the iperf version he/she wants to use
* mention in the documentation what iperf2 versions can be used
* deprecate the support for old iperf version at some point

Typically people do not read the docs after they have a working setup, so from user point of view forcing zephyr to use newer iperf looks like a regression.

But I do not want to hinder development if other devs think there is no major issue here, @rlubos / @pdgendt comments?

rlubos commented:
I would prefer to support a more recent version by default, if needed we can have a Kconfig option enabling legacy support, which could be deprecated over time as suggested.

JordanYates commented:
I am happy to add the choice to revert back to the previous behavior, if someone knows when the header changed format so that the Kconfig description and the docs can be specific.

jukkar commented:
> if someone knows when the header changed format so that the Kconfig description and the docs can be specific.

Definitely iper2 2.0.5 works ok, I think the format changed couple of releases later but I have no concrete release number to give you. I think we can say 2.0.5 here to be in safe side.

JordanYates commented:
> Definitely iper2 2.0.5 works ok, I think the format changed couple of releases later but I have no concrete release number to give you. I think we can say 2.0.5 here to be in safe side.

I can agree that `2.0.5` would use the previous layout, but I don't think that is a particularly useful version number to specify, since as far as I can tell you literally can't download it anymore. At least from https://sourceforge.net/projects/iperf2/files/, which I believe is the official source. Without a concrete version, what are the docs supposed to say?
```
config NET_ZPERF_LEGACY_HEADER
   bool "Legacy zperf UDP header format"
   help
     At some point between version 2.0.5 (2014) and 2.0.14 (2019) the iperf
     UDP header format changed in an incompatible manner that cannot be
     automatically detected. Select this option to use the legacy header format.
```

jukkar commented:
> Without a concrete version, what are the docs supposed to say?

I can try different iperf releases if you really need this info, just a moment.

JordanYates commented:
> > Without a concrete version, what are the docs supposed to say?
> 
> I can try different iperf releases if you really need this info, just a moment.

I don't need it if a vague description is sufficient for the Kconfig

jukkar commented:
2.0.9 works ok, but 2.0.10 did not

JordanYates committed:
doc: zbus: update iperf version information

Update iperf version information for use with zperf. Note the fallback
kconfig option for versions older than `2.0.10`, and remove the warning
about `LAST PACKET NOT RECEIVED!!!` since it is not longer present with
the fixed UDP header.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Updated with the fallback kconfig option and doc updates. The warning is no longer relevant since the library works properly with newer versions now (server stats working as well).
![image](https://github.com/user-attachments/assets/647511c1-49c7-4522-b957-cf3513e4abd6)

jukkar commented on subsys/net/lib/zperf/Kconfig:
```diff
@@ -13,6 +13,14 @@ menuconfig NET_ZPERF
+config NET_ZPERF_LEGACY_HEADER_COMPAT
+	bool "Legacy iperf UDP header format"
+	help
+	  Between iperf 2.0.10 (2017) updated the UDP header format in a
```
Was the `Between ` word a left over from previous version? The sentence looks a bit weird but as I am not a native English speaker the above could be ok to say.

JordanYates replied:
You are right, I missed this when updating the text with the outcome of your testing. I will update tomorrow.
